File: trunk/src/test/java/net/esper/eql/core/TestResultSetProcessorRowPerGroup.java
package net.esper.eql.core;

import java.util.LinkedList;
import java.util.List;

import junit.framework.TestCase;
import net.esper.collection.Pair;
import net.esper.event.EventBean;
import net.esper.support.bean.SupportBean;
import net.esper.support.eql.SupportAggregationService;
import net.esper.support.eql.SupportExprNodeFactory;
import net.esper.support.eql.SupportSelectExprFactory;
import net.esper.support.eql.SupportStreamTypeSvc1Stream;
import net.esper.support.event.SupportEventAdapterService;
import net.esper.support.event.SupportEventBeanFactory;
import net.esper.eql.core.ResultSetProcessorRowPerGroup;
import net.esper.eql.core.SelectExprEvalProcessor;
import net.esper.eql.core.SelectExprProcessor;
import net.esper.eql.expression.ExprNode;

public class TestResultSetProcessorRowPerGroup extends TestCase
{
    private ResultSetProcessorRowPerGroup processor;
    private SupportAggregationService supportAggregationService;

    public void setUp() throws Exception
    {
        SelectExprProcessor selectProcessor = new SelectExprEvalProcessor(SupportSelectExprFactory.makeSelectListFromIdent("string", "s0"),
        		null, false, new SupportStreamTypeSvc1Stream(), SupportEventAdapterService.getService());
        supportAggregationService = new SupportAggregationService();

        List<ExprNode> groupKeyNodes = new LinkedList<ExprNode>();
        groupKeyNodes.add(SupportExprNodeFactory.makeIdentNode("intPrimitive", "s0"));
        groupKeyNodes.add(SupportExprNodeFactory.makeIdentNode("intBoxed", "s0"));

        processor = new ResultSetProcessorRowPerGroup(selectProcessor, null, supportAggregationService, groupKeyNodes, null, false, false);
    }

    public void testProcess()
    {
        EventBean[] newData = new EventBean[] {makeEvent(1, 2), makeEvent(3, 4)};
        EventBean[] oldData = new EventBean[] {makeEvent(1, 2), makeEvent(1, 10)};

        Pair<EventBean[], EventBean[]> result = processor.processViewResult(newData, oldData);

        assertEquals(2, supportAggregationService.getEnterList().size());
        assertEquals(2, supportAggregationService.getLeaveList().size());

        assertEquals(3, result.getFirst().length);
        assertEquals(3, result.getSecond().length);
    }

    private EventBean makeEvent(int intPrimitive, int intBoxed)
    {
        SupportBean bean = new SupportBean();
        bean.setIntPrimitive(intPrimitive);
        bean.setIntBoxed(intBoxed);
        return SupportEventBeanFactory.createObject(bean);
    }
}

File: NEsper.Tests/eql/core/TestResultSetProcessorRowPerGroup.cs
///////////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
///////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;

using NUnit.Framework;

using net.esper.collection;
using net.esper.eql.core;
using net.esper.eql.expression;
using net.esper.events;
using net.esper.support.bean;
using net.esper.support.eql;
using net.esper.support.events;

namespace net.esper.eql.core
{
	[TestFixture]
	public class TestResultSetProcessorRowPerGroup
	{
	    private ResultSetProcessorRowPerGroup processor;
	    private SupportAggregationService supportAggregationService;

	    [SetUp]
	    public void SetUp()
	    {
	        SelectExprProcessor selectProcessor = new SelectExprEvalProcessor(SupportSelectExprFactory.MakeSelectListFromIdent("string", "s0"),
	        		null, false, new SupportStreamTypeSvc1Stream(), SupportEventAdapterService.GetService());
	        supportAggregationService = new SupportAggregationService();

	        IList<ExprNode> groupKeyNodes = new List<ExprNode>();
	        groupKeyNodes.Add(SupportExprNodeFactory.MakeIdentNode("intPrimitive", "s0"));
	        groupKeyNodes.Add(SupportExprNodeFactory.MakeIdentNode("intBoxed", "s0"));

	        processor = new ResultSetProcessorRowPerGroup(selectProcessor, null, supportAggregationService, groupKeyNodes, null, false, false);
	    }

	    [Test]
	    public void testProcess()
	    {
	        EventBean[] newData = new EventBean[] {MakeEvent(1, 2), MakeEvent(3, 4)};
	        EventBean[] oldData = new EventBean[] {MakeEvent(1, 2), MakeEvent(1, 10)};

	        Pair<EventBean[], EventBean[]> result = processor.ProcessViewResult(newData, oldData);

	        Assert.AreEqual(2, supportAggregationService.EnterList.Count);
	        Assert.AreEqual(2, supportAggregationService.LeaveList.Count);

	        Assert.AreEqual(3, result.First.Length);
	        Assert.AreEqual(3, result.Second.Length);
	    }

	    private EventBean MakeEvent(int intPrimitive, int intBoxed)
	    {
	        SupportBean bean = new SupportBean();
	        bean.SetIntPrimitive(intPrimitive);
	        bean.SetIntBoxed(intBoxed);
	        return SupportEventBeanFactory.CreateObject(bean);
	    }
	}
} // End of namespace
